crates/commitlog/src/commitlog.rs

```diff
@@ -0,0 +1,574 @@
+            if let Err(e) = writer.fsync() {
+                warn!("Failed to fsync segment: {e}");
+            }
+            self.start_new_segment()?
+        } else {
+            writer
+        };
+
+        if let Err(e) = writer.commit() {
+            warn!("Commit failed: {e}");
+            self.start_new_segment()?;
+            Err(e)
+        } else {
+            Ok(sz)
+        }
+    }
+
+    pub fn sync(&self) -> io::Result<()> {
+        self.head.fsync()
+    }
+
+    /// The last transaction offset written to disk, or `None` if nothing has
+    /// been written yet.
+    ///
+    /// Note that this does not imply durability: [`Self::sync`] may not have
+    /// been called at this offset.
+    pub fn max_committed_offset(&self) -> Option<u64> {
+        // Naming is hard: the segment's `next_tx_offset` indicates how many
+        // txs are already in the log (it's the next commit's min-tx-offset).
+        // If the value is zero, however, the initial commit hasn't been
+        // committed yet.
+        let next = self.head.next_tx_offset();
+        if next == 0 {
+            None
+        } else {
+            Some(next)
+        }
+    }
+
+    // Helper to obtain a list of the segment offsets which include transaction
+    // offset `offset`.
+    //
+    // The returned `Vec` is sorted in **ascending** order, such that the first
+    // element is the segment which contains `offset`.
+    //
+    // The offset of `self.head` is always included, regardless of how many
+    // entries it actually contains.
+    fn segment_offsets_from(&self, offset: u64) -> Vec<u64> {
+        if offset >= self.head.min_tx_offset {
+            vec![self.head.min_tx_offset]
+        } else {
+            let mut offs = Vec::with_capacity(self.tail.len() + 1);
+            if let Some(pos) = self.tail.iter().rposition(|off| off <= &offset) {
+                offs.extend_from_slice(&self.tail[pos..]);
+                offs.push(self.head.min_tx_offset);
+            }
+
+            offs
+        }
+    }
+
+    pub fn commits_from(&self, offset: u64) -> Commits<R> {
+        let offsets = self.segment_offsets_from(offset);
+        let last_offset = offsets.first().cloned().unwrap_or(offset);
+        let segments = Segments {
+            offs: offsets.into_iter(),
+            repo: self.repo.clone(),
+            max_log_format_version: self.opts.log_format_version,
+        };
+        Commits {
+            inner: None,
+            segments,
+            last_offset,
+            last_error: None,
+        }
+    }
+
+    pub fn reset(mut self) -> io::Result<Self> {
+        info!("hard reset");
+
+        self.tail.reserve(1);
+        self.tail.push(self.head.min_tx_offset);
+        for segment in self.tail.iter().rev() {
+            self.repo.remove_segment(*segment)?;
+        }
+
+        Self::open(self.repo.clone(), self.opts)
+    }
+
+    pub fn reset_to(mut self, offset: u64) -> io::Result<Self> {
```

Comment:
I'm deliberately not documenting private items which either

- Are trivial ("method foo makes a foo")
- Are documented as part of the API, and so would just repeat the text there
- Are merely "helpers" with no particular invariants to document

If you feel otherwise, for example because it aids your understanding, then please submit suggestions. Note also that I'll ask you to motivate why docstrings are needed in those instances.

Reply:
I'll resolve all "Docs :)" comments accordingly.